compiler/src/dotty/tools/dotc/reporting/HideNonSensicalMessages.scala

```diff
@@ -13,7 +13,18 @@ trait HideNonSensicalMessages extends Reporter {
    */
   override def isHidden(dia: Diagnostic)(using Context): Boolean =
     super.isHidden(dia) || {
-        dia.msg.isNonSensical &&
+        (if !errorsReported && dia.isInstanceOf[Diagnostic.Error] then
+          // Bump up errorCount so hasErrors is true while forcing the message.
+          // We use errorsReported as a predicate for broken code.
+          // So now any forcing won't cause, for instance,
+          // assertion errors and thus compiler crashes.
+          // Some messages, once forced, run more code
+          // to generate useful hints for the user.
+          try
+            _errorCount += 1
+            dia.msg.isNonSensical
+          finally _errorCount -= 1 // decrease rather than reset the value so we only ever decrease by 1
+        else dia.msg.isNonSensical) &&
```

Comment:
Why don't we simply test `hasErrors` first? Then the `isNonSensical` test is assured to run with `errorCount > 0`.

Reply:
Good idea, but I had to make another change too.